Phaseme
======

PhaseMe is a tool set to assess the quality of the per read phasing information and help to reduce the errors during this process.


## Input data

1. You require your VCF file to be read based phased, which can be generated by e.g. Whatshap. 
2. 1000 Genomes reference panel haplotypes, which can be obtained from [here](https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.html)
3. You will need to run shapeit similar as below. 


Note that shapeit will complain about ungenotyped variations. Thus we recommend to remove these from the VCF file.


## Step 1: Obtain population information over shapeit


```

shapeit -check --input-vcf  your_out.vcf -R data/1000GP_Phase3/1000GP_Phase3_chr${chr}.hap.gz  data/1000GP_Phase3/1000GP_Phase3_chr${chr}.legend.gz  data/1000GP_Phase3/1000GP_Phase3.sample   --output-log out

shapeit --input-vcf out.vcf  -R data/1000GP_Phase3/1000GP_Phase3_chr${chr}.hap.gz  data/1000GP_Phase3/1000GP_Phase3_chr${chr}.legend.gz  data/1000GP_Phase3/1000GP_Phase3.sample  -M data/1000GP_Phase3/genetic_map_chr${chr}_combined_b37.txt --output-log out1 --output-graph out.graph --exclude-snp  out.snp.strand.exclude

```



## Step 2: Run PhaseMe to obtain stats and improve the quality of phase blocks

```

python2 utilities/samplehaps.py out 500 >log_samplehaps
python utilities/encoderead.py out.hapsamples
python utilities/qc.py out.vcf pairs.txt
python utilities/improve.py out.vcf report.txt
```

## Future steps

The package contains two parts.


1- Reporting quality of phased VCF (including estimated haplotypes).

2- Improving the phased VCF 


