{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "from sys import argv\n",
    "\n",
    "\n",
    "\n",
    "tresh_match_mismatch= 2.2\n",
    "    \n",
    "NEIGHBOURS = 10\n",
    "\n",
    "def read_vcf_file_with10x(vcf_file_address):    \n",
    "    \n",
    "    \"\"\"\n",
    "    Reading the vcf file\n",
    "    \n",
    "    input: vcf file\n",
    "    outputs: \n",
    "            lines_list: list of string. each string is a line of phased vcf file.\n",
    "            var_pos_het_list: genomic position of phased hetrozygous variants in the vcf file. \n",
    "            line_number_het_list: list of line numbers in the vcf file that are phased hetrozygous variant (needed in phasing)\n",
    "            id_blocks: list of ids of phas blocks\n",
    "            allele_blocks: list of list\n",
    "            var_pos_blocks: list of list\n",
    "            stats_vcf: [homozygous0_num, homozygous1_num, hetrozygous_nonphased, hetrozygous_phased, genomic_length_blocks, n50]\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    allele_10x_dic = {}\n",
    "    \n",
    "    vcf_file = open(vcf_file_address,'r')\n",
    "\n",
    "    lines_list=[]                      #  lines  of phased vcf.  needed for reporting improved VCF \n",
    "    var_pos_het_list=[]                # position of phased hetrozygous variants all blocks consequently. \n",
    "\n",
    "    # The followings are for phased hetrozygous variants.\n",
    "    id_blocks = []                 # list of list. Outer list corresponds to phase block. Inner list contains block_id\n",
    "    \n",
    "    allele_blocks = []             # list of list. Outer list corresponds to phase block. Inner list contains alleles of hetro variants \n",
    "    var_pos_blocks = []            # list of list. Outer list corresponds to phase block. Inner list contains genomic positions of hetro variants   \n",
    "\n",
    "    \n",
    "    line_number_het_list = []            # line number of phased hetrozygous variant. We need it for reporting improved version\n",
    "    \n",
    "    first_het_variant = True\n",
    "    line_number = 0\n",
    "    \n",
    "    homozygous0_num = 0\n",
    "    homozygous1_num = 0\n",
    "    hetrozygous_nonphased = 0\n",
    "    hetrozygous_phased = 0\n",
    "    \n",
    "    \n",
    "    for line in vcf_file:\n",
    "        \n",
    "        line_number += 1\n",
    "        \n",
    "        line_strip = line.strip() \n",
    "        \n",
    "        lines_list.append(line_strip)\n",
    "        \n",
    "        if line_strip.startswith('#'):\n",
    "            pass \n",
    "            #header_lines_list.append(line_strip)\n",
    "            #sample_names = line_strip.split('\\t')[9:11]            # last line of header contains sample name\n",
    "        else:\n",
    "\n",
    "            line_parts = line_strip.split('\\t') \n",
    "\n",
    "            chrom = line_parts[0]\n",
    "\n",
    "            #if str(chrom_output)!=chrom:\n",
    "                #print(chrom_output,chrom)\n",
    "\n",
    "            var_pos = int(line_parts[1])                           # genomic position of variants\n",
    "            \n",
    "\n",
    "            \n",
    "            format_genotype, values_genotype = line_parts[8:10]    # 'GT:GQ:DP:AF:GL:PS', '0|1:255:.:.:.,0,.:60780'\n",
    "            \n",
    "            values_genotype_splitted = values_genotype.split(':')\n",
    "            format_genotype_splitted = format_genotype.split(':')\n",
    "            \n",
    "            gt_index = format_genotype_splitted.index(\"GT\")           #  index of allele in  values_genotype \n",
    "            \n",
    "            allele = values_genotype_splitted[gt_index]\n",
    "            \n",
    "            \n",
    "            # how should we handle '2' in allele ?  \n",
    "            \n",
    "            if './.' in allele:\n",
    "                print(\"There is a vriant with genomic position \"+str(var_pos)+\" that is not genotyped. Remove it first.\")\n",
    "                exit(1)\n",
    "                \n",
    "            \n",
    "            # if '/' in allele: print(\"There is a vriant with genomic position \"+str(var_pos)+\" that is not phased. Remove it first.\")            \n",
    "\n",
    "            if allele == '0/1' or allele == '1/0':\n",
    "                hetrozygous_nonphased += 1\n",
    "\n",
    "            if allele == '0|0' or allele == '0/0':\n",
    "                homozygous0_num += 1\n",
    "                \n",
    "            if allele == '1|1' or allele == '1/1':\n",
    "                homozygous1_num += 1\n",
    "                    \n",
    "            if (allele == '0|1' or allele == '1|0'):\n",
    "            \n",
    "                hetrozygous_phased += 1\n",
    "                \n",
    "                var_pos_het_list.append(var_pos)\n",
    "                \n",
    "                line_number_het_list.append(line_number)\n",
    "                \n",
    "                ps_index = format_genotype_splitted.index(\"PS\")           #  index of phase set in values_genotype \n",
    "                id_block = values_genotype_splitted[ps_index]\n",
    "                \n",
    "                \n",
    "                \n",
    "                if first_het_variant:           # for the first het variant\n",
    "                    first_het_variant = False\n",
    "                    \n",
    "                    allele_block = [allele]\n",
    "                    var_pos_block = [int(var_pos)]\n",
    "                    id_blocks.append(id_block)\n",
    "                    \n",
    "                    \n",
    "                else:                              # for the rest of het variants\n",
    "                    if id_block in id_blocks:\n",
    "                        allele_block.append(allele)\n",
    "                        var_pos_block.append(int(var_pos)) \n",
    "\n",
    "                    else:\n",
    "\n",
    "                        # add previous block to the list of all blocks\n",
    "                        allele_blocks.append(allele_block)\n",
    "                        var_pos_blocks.append(var_pos_block)\n",
    "\n",
    "                        # creat new phase block\n",
    "                        allele_block = [allele]\n",
    "                        var_pos_block = [int(var_pos)]\n",
    "                        id_blocks.append(id_block)\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "                values_genotype_10x = line_parts[10]\n",
    "                values_genotype_10x_splitted = values_genotype_10x.split(':')\n",
    "                allele_10x = values_genotype_10x_splitted[gt_index]\n",
    "                id_block_10x = values_genotype_10x_splitted[ps_index]\n",
    "                if allele_10x != './.':\n",
    "                    allele_10x_dic[var_pos]= str(id_block_10x)+':'+str(allele_10x)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    # # for the last het variant, we  finish the last block.\n",
    "    allele_blocks.append(allele_block)\n",
    "    var_pos_blocks.append(var_pos_block)\n",
    "    \n",
    "    \n",
    "          \n",
    "    genomic_length_blocks = [] \n",
    "    \n",
    "    for var_pos_block in var_pos_blocks:\n",
    "        genomic_length_blocks.append(var_pos_block[-1]-var_pos_block[0])\n",
    "        \n",
    "    values_sorted = sorted(genomic_length_blocks, reverse=True)\n",
    "    csum = np.cumsum(values_sorted)\n",
    "    \n",
    "    n2 = int(sum(values_sorted)/2)\n",
    "    csumn2 = min(csum[csum >= n2])\n",
    "    ind = np.where(csum == csumn2)\n",
    "    n50 = values_sorted[int(ind[0])]\n",
    "\n",
    "        \n",
    "    \n",
    "    stats_vcf = [homozygous0_num, homozygous1_num, hetrozygous_nonphased, hetrozygous_phased, genomic_length_blocks, n50]\n",
    "\n",
    "\n",
    "\n",
    "    return lines_list, var_pos_het_list, line_number_het_list, id_blocks, allele_blocks, var_pos_blocks, stats_vcf, allele_10x_dic\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#def read_file_mismatches(file_mismatches_address, id_blocks):\n",
    "def read_file_mismatches_with10x(file_mismatches_address, id_blocks):\n",
    "\n",
    "    file = open(file_mismatches_address,'r')\n",
    "        \n",
    "    \n",
    "    first_variant = True\n",
    "    block_ids = []       # this is only for checking consitency between phased VCF and the report file\n",
    "    \n",
    "    \n",
    "    for line in file:\n",
    "        \n",
    "        if line.strip().startswith('#c'): # new chromosome\n",
    "            comparison_result_blocks = []\n",
    "        \n",
    "        elif line.strip().startswith('#P'):  # new block\n",
    "            \n",
    "            if not first_variant:\n",
    "                comparison_result_blocks.append(comparison_result_block)\n",
    "                \n",
    "            comparison_result_block = []\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            first_variant=False\n",
    "            \n",
    "            line_parts_with10x = line.strip().split('\\t')     # ['42081', '0', '42096', '0']\n",
    "            \n",
    "            line_parts = line_parts_with10x[2:]\n",
    "            \n",
    "            chrom = line_parts[0]\n",
    "            var_pos = int(line_parts[1])\n",
    "                        \n",
    "            line_part_2 = line_parts[2].split(':')  # 60780:1|0\n",
    "\n",
    "            block_id = line_part_2[0]\n",
    "            \n",
    "            if block_id not in block_ids:\n",
    "                block_ids.append(block_id)\n",
    "            \n",
    "            allele = line_part_2[1]\n",
    "            \n",
    "            if line_parts[3] == '.':\n",
    "                mismatched_list = []\n",
    "            else:\n",
    "                mismatched_list = line_parts[3].split(',')\n",
    "\n",
    "            if line_parts[4] == '.':\n",
    "                matched_list = []\n",
    "            else:\n",
    "                matched_list = line_parts[4].split(',') \n",
    "                \n",
    "            comparison_result_block.append([matched_list, mismatched_list])\n",
    "                        \n",
    "            block_id_previous = block_id\n",
    "            \n",
    "    # add last block\n",
    "    comparison_result_blocks.append(comparison_result_block)\n",
    "    \n",
    "    if block_ids != id_blocks:\n",
    "        print(\"The block ids of phased VCF are not consitent with that of report file. The phased VCF file should be the same in all steps of package.\\n\")\n",
    "                \n",
    "            \n",
    "                \n",
    "            \n",
    "    return comparison_result_blocks\n",
    "\n",
    "\n",
    "def read_file_pairs_forward(file_pairs_address):\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    Reading the pairs.txt file and save it in a dictinary only once.\n",
    "\n",
    "    input: pairs.txt\n",
    "    Each line of input file has three part \n",
    "        position of var1, position of  var2, realtion between phasing of var1 and that of var2\n",
    "        10 20 identical\n",
    "        \n",
    "        Each pair reports only one in output dictionary. We report the example for the key   20:[[10],[]]\n",
    "        Becuase, in phase block we go variant by vairant. So the phasing of previous variants are important not later's\n",
    "        \n",
    "    output: a dictinary, key: genomic position of variant, value: [[],[]] \n",
    "    \n",
    "            first list: the genomic position of variant with identical phasing \n",
    "            second list: the genomic position of variant with opposite phasing\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    file_pairs = open(file_pairs_address,'r'); \n",
    "    \n",
    "    pop_inf_dic = {}                              # key variant index, value two lists \n",
    "    for line in file_pairs:\n",
    "        \n",
    "        line_parts = line.strip().split('\\t')     # ['42081', '0', '42096', '0']\n",
    "\n",
    "        \n",
    "        \n",
    "        snv1_pos = int(line_parts[0]) \n",
    "        snv2_pos = int(line_parts[1])                \n",
    "\n",
    "        relation_phasing = line_parts[2]\n",
    "        \n",
    "        snvs_pos = [snv1_pos, snv2_pos]\n",
    "        host_pos = max(snvs_pos)\n",
    "        guest_pos = min(snvs_pos)\n",
    "        \n",
    "        if host_pos not in pop_inf_dic.keys():\n",
    "            \n",
    "            # first list identical phase\n",
    "            # second list opposite phase\n",
    "            pop_inf_dic[host_pos] = [[], []]\n",
    "            \n",
    "        if relation_phasing == 'identical':\n",
    "            pop_inf_dic[host_pos][0].append(guest_pos)  \n",
    "        elif relation_phasing == 'opposite':\n",
    "            pop_inf_dic[host_pos][1].append(guest_pos) \n",
    "            \n",
    "    return pop_inf_dic  # [vars_identical_phase, vars_opposite_phase]=pop_inf_dic[var_pos]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def read_file_pairs_forward_backward(file_pairs_address):\n",
    "\n",
    "    \"\"\"\n",
    "    see read_file_pairs_forward\n",
    "    \n",
    "    The only difference is that a pair is reported twic. \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    file_pairs = open(file_pairs_address,'r'); \n",
    "    \n",
    "    pop_inf_dic = {}                              # key variant index, value two lists \n",
    "    for line in file_pairs:\n",
    "        \n",
    "        line_parts = line.strip().split('\\t')     # ['42081', '0', '42096', '0']\n",
    "\n",
    "         \n",
    "        snv1_pos = int(line_parts[0]) \n",
    "        snv2_pos = int(line_parts[1])                \n",
    "\n",
    "        relation_phasing = line_parts[2]\n",
    "        \n",
    "        host_pos = snv1_pos\n",
    "        guest_pos = snv2_pos\n",
    "        \n",
    "        if host_pos not in pop_inf_dic.keys():\n",
    "            \n",
    "            # first list identical phase\n",
    "            # second list opposite phase\n",
    "            pop_inf_dic[host_pos] = [[], []]\n",
    "            \n",
    "        if relation_phasing == 'identical':\n",
    "            pop_inf_dic[host_pos][0].append(guest_pos)  \n",
    "        elif relation_phasing == 'opposite':\n",
    "            pop_inf_dic[host_pos][1].append(guest_pos) \n",
    "   \n",
    "            \n",
    "        host_pos = snv2_pos\n",
    "        guest_pos = snv1_pos\n",
    "        \n",
    "        if host_pos not in pop_inf_dic.keys():\n",
    "            \n",
    "            # first list identical phase\n",
    "            # second list opposite phase\n",
    "            pop_inf_dic[host_pos] = [[], []]\n",
    "            \n",
    "        if relation_phasing == 'identical':\n",
    "            pop_inf_dic[host_pos][0].append(guest_pos)  \n",
    "        elif relation_phasing == 'opposite':\n",
    "            pop_inf_dic[host_pos][1].append(guest_pos) \n",
    "            \n",
    "    return pop_inf_dic  # [vars_identical_phase, vars_opposite_phase]=pop_inf_dic[var_pos]\n",
    "\n",
    "      \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def compare_phase_block_pop(allele_block, var_pos_block, pop_inf_dic, lower_bound, upper_bound):\n",
    "\n",
    "    \n",
    "    \"\"\"\n",
    "    Compare alleles of phased VCF with pairs (from the population information)\n",
    "    \n",
    "    input:  allele_block, var_pos_block    (phased VCF)\n",
    "            pop_inf_dic \n",
    "    \n",
    "    output: a list of [[],[]] \n",
    "    \n",
    "                        Inner list correspond to a variant (host). \n",
    "                        first inner list: positons of variants that are matched with the host variant (based on population information)\n",
    "                                         the status of match can be either identical or opposite phasing.                \n",
    "                        second inner list mismatched\n",
    "    \n",
    "    \n",
    "    Exmpale 1:  pair from population information:  10 20 identical\n",
    "                if phased vcf   10 0|1    20 0|1, the 10 and 20 are matched.\n",
    "                if phased vcf   10 0|1    20 1|0, the 10 and 20 are mismatched.\n",
    "\n",
    "    Exmpale 2:  pair from population information:  10 20 opposite\n",
    "                if phased vcf   10 0|1    20 0|1, the 10 and 20 are mismatched.\n",
    "                if phased vcf   10 0|1    20 1|0, the 10 and 20 are matched.\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    # the results of comparison between \n",
    "    \n",
    "    comparison_result_block = []\n",
    "    for var_i in range(len(allele_block)):\n",
    "\n",
    "        #var_i is the index of variant within the block of phased vcf\n",
    "\n",
    "        var_pos = var_pos_block[var_i]\n",
    "        allele = allele_block[var_i]\n",
    "\n",
    "        if var_pos >= lower_bound and  var_pos <= upper_bound:\n",
    "\n",
    "            #alleles=hap_block[var_i]\n",
    "            #var_idx=idc_block[var_i]\n",
    "            #var_i is the index of variant within block\n",
    "            #var_idx is the index of vriant globally in the VCF file \n",
    "            if var_pos in pop_inf_dic.keys():\n",
    "    \n",
    "                [vars_identical_phase, vars_opposite_phase] = pop_inf_dic[var_pos]  \n",
    "        \n",
    "                \n",
    "                matched_identical_phase_list = []\n",
    "                mismatched_identical_phase_list = []\n",
    "                matched_opposite_phase_list = []\n",
    "                mismatched_opposite_phase_list = []\n",
    "                #for sim_idx in vars_identical_phase: # sim shows the relation between two elements of a pop pair\n",
    "                \n",
    "                \n",
    "                # The differene between the two following for is the comparing the allele_var_guest and  allele (flliped)\n",
    "                \n",
    "                for var_pos_identical_phase in vars_identical_phase: # Those variant that have the same phasing with var_pos\n",
    "\n",
    "                    if var_pos_identical_phase >= lower_bound and var_pos_identical_phase <= upper_bound:\n",
    "                        try:         # if the SNP  of pairs is in this block of phased VCF\n",
    "                            \n",
    "                            allele_var_guest= allele_block[var_pos_block.index(var_pos_identical_phase)]\n",
    "                                                   \n",
    "                            if allele_var_guest == allele: \n",
    "                                matched_identical_phase_list.append(var_pos_identical_phase)\n",
    "                            else:\n",
    "                                mismatched_identical_phase_list.append(var_pos_identical_phase) \n",
    "\n",
    "                        except:\n",
    "                            pass\n",
    "\n",
    "                    \n",
    "                for var_pos_opposite_phase in vars_opposite_phase: # Those variant that have the same phasing with var_pos\n",
    "                    \n",
    "                    if var_pos_opposite_phase >= lower_bound and  var_pos_opposite_phase <= upper_bound:\n",
    "                        try:         # if the SNP  of pairs is in this block of phased VCF        \n",
    "                            allele_var_guest = allele_block[var_pos_block.index(var_pos_opposite_phase)]\n",
    "                                                \n",
    "                            if allele_var_guest == str(1-int(allele[0]))+'|'+str(1-int(allele[2])):                   \n",
    "                                matched_opposite_phase_list.append(var_pos_opposite_phase)\n",
    "                            else:\n",
    "                                mismatched_opposite_phase_list.append(var_pos_opposite_phase) \n",
    "                        except:\n",
    "                            pass                \n",
    "                    \n",
    "                matched_list = matched_identical_phase_list + matched_opposite_phase_list\n",
    "                mismatched_list = mismatched_identical_phase_list + mismatched_opposite_phase_list\n",
    "                \n",
    "                comparison_result = [matched_list, mismatched_list]\n",
    "            else:\n",
    "                comparison_result = [[],[]]\n",
    "        else:\n",
    "            comparison_result = [[],[]]\n",
    "                \n",
    "        comparison_result_block.append(comparison_result)\n",
    "                \n",
    "        #print('comparison is done')\n",
    "    return  comparison_result_block\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def decide_flip_cut(id_blocks, allele_blocks, var_pos_blocks, comparison_result_blocks):\n",
    "    \n",
    "    flip_list = []\n",
    "    cut_list_blocks = []\n",
    "\n",
    "    for block_i, block_id  in enumerate(id_blocks):\n",
    "\n",
    "        allele_block = allele_blocks[block_i]       # hap_block\n",
    "        var_pos_block = var_pos_blocks[block_i] \n",
    "        \n",
    "        \n",
    "        # after deciding the cut, it is not applied, so the number of block is not changed.\n",
    "        # but after deciding the match/mismatch information is updated for those varaint afterwards and removed for previose since new block is started practically. \n",
    "        \n",
    "        comparison_result_block = comparison_result_blocks[block_i] # ont_pop_block\n",
    "\n",
    "        \n",
    "        cut_list_block=[]\n",
    "\n",
    "        for var_i, var_pos in enumerate(var_pos_block):  # var_pos_block idc_block\n",
    "            \n",
    "            # var_i is the index of variant within block\n",
    "            # var_pos is the genomic position of variant\n",
    "\n",
    "            \n",
    "            \n",
    "            comparison_result = comparison_result_block[var_i]  # rightafter we decide a cut, we use an updated version of comparison_result_block\n",
    "\n",
    "            if comparison_result == [[],[]]:  # there is no population information for this variant\n",
    "                out_pairs = '.'\n",
    "                \n",
    "            else:\n",
    "                [matched_list, mismatched_list] = comparison_result    \n",
    "                \n",
    "                num_mismatched = len(mismatched_list)\n",
    "                num_matched = len(matched_list)\n",
    "                \n",
    "                condition_flip_cut = 0\n",
    "                condition_flip = 0\n",
    "                condition_cut = 0\n",
    "                \n",
    "                if num_mismatched >= 2:\n",
    "\n",
    "                    [matched_next1, mismatched_next1] = comparison_result_block[var_i+1]\n",
    "                    [matched_next2, mismatched_next2] = comparison_result_block[var_i+2]\n",
    "\n",
    "                    num_matched_next1 = len(matched_next1)\n",
    "                    num_mismatched_next1 = len(mismatched_next1)\n",
    "                    \n",
    "                    \n",
    "                    if num_matched == 0 or num_matched/num_mismatched < tresh_match_mismatch:\n",
    "                        if num_mismatched_next1 >=1 and (num_matched_next1 == 0 or num_matched_next1/num_mismatched_next1 < tresh_match_mismatch):\n",
    "                            condition_cut = 1\n",
    "                    \n",
    "                    if var_pos in mismatched_next1 and var_pos in mismatched_next2:\n",
    "                        condition_flip = 1\n",
    "\n",
    "                        \n",
    "                if condition_flip:\n",
    "                    print('flip candidate', var_pos)\n",
    "                    flip_list.append(var_pos)\n",
    "                    \n",
    "                if condition_cut and not condition_flip:\n",
    "                    cut_pos = var_pos                 # cut_pos the starting position of new block\n",
    "                    print('cut candidate', cut_pos) \n",
    "                    \n",
    "                    cut_list_block.append(var_pos)\n",
    "                    \n",
    "                    lower_bound = var_pos\n",
    "                    upper_bound = var_pos_block[-1]\n",
    "                    \n",
    "                    # after deciding the cut, it is not applied, so the number of block is not changed.\n",
    "                    # but after deciding the match/mismatch information is updated for those varaint afterwards and removed for previose since new block is started practically. \n",
    "        \n",
    "                    comparison_result_block = compare_phase_block_pop(allele_block, var_pos_block, pop_inf_dic, lower_bound, upper_bound)\n",
    "        cut_list_blocks.append(cut_list_block)\n",
    "    return flip_list, cut_list_blocks             \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def improve_vcf_flip(lines_list, line_number_het_list, flip_list):\n",
    "    \n",
    "    \n",
    "    # becareful about the concpet of deep copy. the change will affect the lines_list\n",
    "    lines_list_improved_flipping = lines_list\n",
    "    \n",
    "    for var_pos_flip in flip_list:\n",
    "        \n",
    "       \n",
    "        var_index_flip = var_pos_het_list.index(var_pos_flip)\n",
    "        \n",
    "        var_line_number_flip = line_number_het_list[var_index_flip]    # 1-based line number \n",
    "                \n",
    "        line = lines_list[var_line_number_flip-1]                      # 0-based  list\n",
    "\n",
    "        line_parts = line.split('\\t')\n",
    "\n",
    "        format_genotype, values_genotype = line_parts[8:10]    # 'GT:GQ:DP:AF:GL:PS', '0|1:255:.:.:.,0,.:60780'\n",
    "\n",
    "        values_genotype_splitted = values_genotype.split(':')\n",
    "        format_genotype_splitted = format_genotype.split(':')\n",
    "\n",
    "        gt_index = format_genotype_splitted.index(\"GT\")           #  index of allele in  values_genotype \n",
    "        allele = values_genotype_splitted[gt_index]\n",
    "        \n",
    "        allele_flipped = str(1-int(allele[0]))+'|'+str(1-int(allele[2]))\n",
    "        values_genotype_splitted[gt_index] = allele_flipped\n",
    "        \n",
    "        line_parts[9]=':'.join(values_genotype_splitted)\n",
    "        \n",
    "        line = '\\t'.join(line_parts)\n",
    "        \n",
    "        lines_list_improved_flipping[var_line_number_flip-1] = line\n",
    "\n",
    "\n",
    "        \n",
    "    return lines_list_improved_flipping  \n",
    "   \n",
    "    \n",
    "\n",
    "\n",
    "def improve_vcf_cut(lines_list_improved_flipping, id_blocks, cut_list_blocks, var_pos_blocks):\n",
    "\n",
    "\n",
    "    var_blockid_dic_updated = {}   # key: var_pos (genomic position of variant) value: block_id   after enforcing cuts!\n",
    "    \n",
    "    \n",
    "    for block_i, block_id in enumerate(id_blocks):\n",
    "        \n",
    "        cut_list_block = cut_list_blocks[block_i]\n",
    "        var_pos_block = var_pos_blocks[block_i]\n",
    "        \n",
    "        boundries_list = [var_pos_block[0]]+cut_list_block+[var_pos_block[-1]]   # including start, end, and the cut list of  each block\n",
    "        \n",
    "        for i in range(1,len(boundries_list)):\n",
    "            \n",
    "            prev_cut_pos = boundries_list[i-1]\n",
    "            next_cut_pos = boundries_list[i]\n",
    "            \n",
    "            blockid = prev_cut_pos\n",
    "            \n",
    "            for var_pos in var_pos_block:\n",
    "                \n",
    "                if var_pos >= prev_cut_pos and var_pos < next_cut_pos:\n",
    "                    var_blockid_dic_updated[var_pos] = str(prev_cut_pos)\n",
    "        # for last variant in the block\n",
    "        var_blockid_dic_updated[boundries_list[i]] = str(boundries_list[i-1])\n",
    "        \n",
    "    \n",
    "    lines_list_improved_cut = lines_list_improved_flipping\n",
    "\n",
    "    for var_i in range(len(var_pos_het_list)): \n",
    "        \n",
    "        var_pos = var_pos_het_list[var_i]\n",
    "        line_number = line_number_het_list[var_i]   # 1-based line number \n",
    "         \n",
    "        line = lines_list_improved_cut[line_number-1]            # 0-based  list\n",
    "         \n",
    "        line_parts = line.split('\\t')\n",
    "\n",
    "        format_genotype, values_genotype = line_parts[8:10]    # 'GT:GQ:DP:AF:GL:PS', '0|1:255:.:.:.,0,.:60780'\n",
    "\n",
    "        values_genotype_splitted = values_genotype.split(':')\n",
    "        format_genotype_splitted = format_genotype.split(':')\n",
    "\n",
    "        gt_index = format_genotype_splitted.index(\"GT\")           #  index of allele in  values_genotype\n",
    "        ps_index = format_genotype_splitted.index(\"PS\")\n",
    "\n",
    "        allele = values_genotype_splitted[gt_index]\n",
    "\n",
    "        block_id_updated = var_blockid_dic_updated[var_pos]\n",
    "\n",
    "        values_genotype_splitted[ps_index] = block_id_updated\n",
    "\n",
    "        line_parts[9] = ':'.join(values_genotype_splitted)\n",
    "\n",
    "        lines_list_improved_cut[line_number-1]  = '\\t'.join(line_parts)\n",
    "\n",
    "\n",
    "    return lines_list_improved_cut\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "def write_out_vcf(lines_list, lines_list_improved_cut):\n",
    "\n",
    "    vcf_file_improved = open(vcf_file_improved_address,'w');  # phased_vcf_dic\n",
    "\n",
    "    for header_line in lines_list_improved_cut:\n",
    "        vcf_file_improved.write(header_line+'\\n')\n",
    "        \n",
    "\n",
    "    vcf_file_improved.close()\n",
    "    \n",
    "    return 1 \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    \n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    Input:\n",
    "    \n",
    "    Output:\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    chrom = 22\n",
    "    vcf_file_address = 'input/'+str(chrom)+'_ont_10x.vcf'\n",
    "    file_mismatches_address = 'Expected_output/'+str(chrom)+'_report_mismatches_both.txt'\n",
    "    file_pairs_address = 'input/'+str(chrom)+'_pairs.txt'\n",
    "\n",
    "\n",
    "\n",
    "                \n",
    "    lines_list, var_pos_het_list, line_number_het_list, id_blocks, allele_blocks, var_pos_blocks, stats_vcf, allele_10x_dic  =  read_vcf_file_with10x(vcf_file_address)\n",
    "\n",
    "    comparison_result_blocks = read_file_mismatches_with10x(file_mismatches_address, id_blocks)\n",
    "    \n",
    "    \n",
    "    \n",
    "    Forward_or_both = 'forward'\n",
    "    \n",
    "    if Forward_or_both == 'forward':\n",
    "        pop_inf_dic = read_file_pairs_forward(file_pairs_address)         \n",
    "        \n",
    "    if Forward_or_both == 'both':                       # reproting pairs both forward and backward\n",
    "        pop_inf_dic = read_file_pairs_forward_backward(file_pairs_address)    \n",
    "        \n",
    "        \n",
    "\n",
    "#     flip_list, cut_list_blocks = decide_flip_cut(id_blocks, allele_blocks, var_pos_blocks, comparison_result_blocks)\n",
    "\n",
    "#     # becareful about the concpet of deep copy.  Changes may affect the lines_list\n",
    "#     lines_list_improved_flipping = improve_vcf_flip(lines_list, line_number_het_list, flip_list)  # it may contain homo vars\n",
    "    \n",
    "    \n",
    "#     lines_list_improved_cut= improve_vcf_cut(lines_list_improved_flipping, id_blocks, cut_list_blocks, var_pos_blocks)\n",
    "\n",
    "#     vcf_file_improved_address = ''+str(chrom)+'_ont_10x_improved.vcf' \n",
    "#     write_out_vcf(vcf_file_improved_address, lines_list_improved_cut)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on block : 16052419\n",
      "flip candidate 16471042\n",
      "working on block : 16847870\n",
      "flip candidate 16951562\n",
      "working on block : 18712732\n",
      "flip candidate 18873617\n",
      "working on block : 20610824\n",
      "flip candidate 21731913\n",
      "flip candidate 29970272\n",
      "flip candidate 39053382\n",
      "flip candidate 45013019\n",
      "flip candidate 45304914\n",
      "flip candidate 45454275\n",
      "flip candidate 46136599\n",
      "flip candidate 49437557\n",
      "working on block : 50415657\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if 1: #def decide_flip_cut(id_blocks, allele_blocks, var_pos_blocks, comparison_result_blocks):\n",
    "    \n",
    "    flip_list = []\n",
    "    cut_list_blocks = []\n",
    "    \n",
    "    block_i = 0 \n",
    "    block_id = id_blocks[block_i]\n",
    "    \n",
    "    \n",
    "    \n",
    "    for block_i, block_id  in enumerate(id_blocks):\n",
    "\n",
    "        \n",
    "        print('working on block :',block_id)\n",
    "        allele_block = allele_blocks[block_i]       # hap_block\n",
    "        var_pos_block = var_pos_blocks[block_i] \n",
    "        \n",
    "        \n",
    "        # after deciding the cut, it is not applied, so the number of block is not changed.\n",
    "        # but after deciding the match/mismatch information is updated for those varaint afterwards and removed for previose since new block is started practically. \n",
    "        \n",
    "        comparison_result_block = comparison_result_blocks[block_i] # ont_pop_block\n",
    "\n",
    "        \n",
    "        cut_list_block=[]\n",
    "\n",
    "        for var_i, var_pos in enumerate(var_pos_block):  # var_pos_block idc_block\n",
    "            \n",
    "            # var_i is the index of variant within block\n",
    "            # var_pos is the genomic position of variant\n",
    "\n",
    "            \n",
    "            \n",
    "            comparison_result = comparison_result_block[var_i]  # rightafter we decide a cut, we use an updated version of comparison_result_block\n",
    "\n",
    "            if comparison_result == [[],[]]:  # there is no population information for this variant\n",
    "                out_pairs = '.'\n",
    "                \n",
    "            else:\n",
    "                [matched_list, mismatched_list] = comparison_result    \n",
    "                \n",
    "                num_mismatched = len(mismatched_list)\n",
    "                num_matched = len(matched_list)\n",
    "                \n",
    "                condition_flip_cut = 0\n",
    "                condition_flip = 0\n",
    "                condition_cut = 0\n",
    "                \n",
    "                if num_mismatched >= 2:\n",
    "                    \n",
    "                    [matched_next1, mismatched_next1] = comparison_result_block[var_i+1]\n",
    "                    [matched_next2, mismatched_next2] = comparison_result_block[var_i+2]\n",
    "                    \n",
    "                    \n",
    "                    num_matched_next1 = len(matched_next1)\n",
    "                    num_mismatched_next1 = len(mismatched_next1)\n",
    "                    \n",
    "                    \n",
    "#                     if num_matched == 0 or num_matched/num_mismatched < tresh_match_mismatch:\n",
    "#                         if num_mismatched_next1 >=1 and (num_matched_next1 == 0 or num_matched_next1/num_mismatched_next1 < tresh_match_mismatch):\n",
    "#                             condition_cut = 1\n",
    "                    \n",
    "                    if var_pos in mismatched_next1 and var_pos in mismatched_next2:\n",
    "                        condition_flip = 1\n",
    "\n",
    "                        \n",
    "                    pairs_subseqeunt_neighbours_all = 0\n",
    "                    pairs_previous_neighbours_all = 0\n",
    "                    \n",
    "                    num_membership_mis_subseqeunt = 0\n",
    "                    num_membership_mis_previous = 0\n",
    "                    \n",
    "                    for var_neighbour_j in range(var_i+1, min(len(var_pos_block),var_i+NEIGHBOURS+1)):\n",
    "                        [matched_neighbour_j, mismatched_neighbour_j] = comparison_result_block[var_neighbour_j]\n",
    "                        \n",
    "                        pairs_subseqeunt_neighbours_all += len(matched_neighbour_j)+len(mismatched_neighbour_j)\n",
    "                        \n",
    "                        if str(var_pos) in mismatched_neighbour_j:\n",
    "                            num_membership_mis_subseqeunt +=1\n",
    "                    \n",
    "                    \n",
    "                    for var_neighbour_j in range(max(0,var_i-NEIGHBOURS-1),var_i):\n",
    "                        [matched_neighbour_j, mismatched_neighbour_j] = comparison_result_block[var_neighbour_j]\n",
    "                        \n",
    "                        pairs_previous_neighbours_all += len(matched_neighbour_j)+len(mismatched_neighbour_j)\n",
    "                        \n",
    "                        if str(var_pos) in mismatched_neighbour_j:\n",
    "                            num_membership_mis_previous +=1\n",
    "                    \n",
    "                    #print(var_pos,num_membership_mis)\n",
    "\n",
    "                    \n",
    "                    \n",
    "                    if  num_membership_mis_subseqeunt >= 1 and num_membership_mis_previous >= 1:\n",
    "                        condition_flip = 1 \n",
    "\n",
    "                    if  num_membership_mis_subseqeunt >= 2 and (pairs_previous_neighbours_all == 0):   \n",
    "                        condition_flip = 1 \n",
    "                        \n",
    "                    if  num_membership_mis_previous >= 2 and (pairs_subseqeunt_neighbours_all == 0):\n",
    "                        condition_flip = 1 \n",
    "\n",
    "                    \n",
    "                        \n",
    "                if condition_flip:\n",
    "                    print('flip candidate', var_pos)\n",
    "                    flip_list.append(var_pos)\n",
    "                    \n",
    "#                 if condition_cut :#and not condition_flip:\n",
    "#                     cut_pos = var_pos                 # cut_pos the starting position of new block\n",
    "#                     print('cut candidate', cut_pos) \n",
    "                    \n",
    "#                     cut_list_block.append(var_pos)\n",
    "                    \n",
    "                    lower_bound = var_pos\n",
    "                    upper_bound = var_pos_block[-1]\n",
    "                    \n",
    "                    # after deciding the cut, it is not applied, so the number of block is not changed.\n",
    "                    # but after deciding the match/mismatch information is updated for those varaint afterwards and removed for previose since new block is started practically. \n",
    "        \n",
    "                    comparison_result_block = compare_phase_block_pop(allele_block, var_pos_block, pop_inf_dic, lower_bound, upper_bound)\n",
    "        cut_list_blocks.append(cut_list_block)\n",
    "   # return flip_list, cut_list_blocks             \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flip_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[16471042,\n",
       " 16951562,\n",
       " 21731913,\n",
       " 29970272,\n",
       " 39053382,\n",
       " 45013019,\n",
       " 45304914,\n",
       " 45454275,\n",
       " 46136599,\n",
       " 49437557]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# both\n",
    "#flip_list.pop(2)\n",
    "flip_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[16471042,\n",
       " 16951562,\n",
       " 18873617,\n",
       " 21731913,\n",
       " 29970272,\n",
       " 39053382,\n",
       " 45013019,\n",
       " 45304914,\n",
       " 45454275,\n",
       " 46136599,\n",
       " 49437557]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@ both\n",
    "flip_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cut_list_blocks=a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# forward\n",
    "#cut_list_blocks\n",
    "\n",
    "a= [[],\n",
    " [16938155, 16960113],\n",
    " [18873806],\n",
    " [21731913,\n",
    "  24916182,\n",
    "  27896908,\n",
    "  27904170,\n",
    "  37400787,\n",
    "  39053382,\n",
    "  43134801,\n",
    "  44814180,\n",
    "  46247222],\n",
    " []]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cut_list_blocks = a\n",
    "\n",
    "flip_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines_list_improved_flipping = improve_vcf_flip(lines_list, line_number_het_list, flip_list)  # it may contain homo vars\n",
    "\n",
    "\n",
    "lines_list_improved_cut= improve_vcf_cut(lines_list_improved_flipping, id_blocks, cut_list_blocks, var_pos_blocks)\n",
    "\n",
    "vcf_file_improved_address = ''+str(chrom)+'_ont_10x_improved_2.2_flipFrw_b.vcf' \n",
    "write_out_vcf(vcf_file_improved_address, lines_list_improved_cut)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
