{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of blocks in ont 22\n",
      "63828 22109\n",
      "18098887 11005\n",
      "27607877 11132\n",
      "38869204 3743\n",
      "42354942 1293\n",
      "42596699 5413\n",
      "46486766 106\n",
      "46666247 181\n",
      "46787270 1854\n",
      "47532101 413\n",
      "47894257 270\n",
      "48108270 1137\n",
      "49088577 8\n",
      "49196518 1785\n",
      "51215194 172\n",
      "51449190 389\n",
      "51722995 68171\n",
      "125922555 3004\n",
      "128766436 6149\n",
      "133431868 342\n",
      "133730171 507\n",
      "133987652 2598\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['1|0', '1|0']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grep -e \"PATMAT\" -e \"#\" 10x22.vcf | grep -v \"0/1\" > 10x22_ptmt.vcf\n",
    "import numpy as np\n",
    "chrom_output=10\n",
    "vcf_file_address = 'data/'+str(chrom_output)+'/out.vcf' #tst.vcf' #\n",
    "vcf_file = open(vcf_file_address,'r')\n",
    "\n",
    "hap_blocks={} # key: id of phase block (i.e. phase st), value=allele_1 \n",
    "idc_blocks={} # key: id of phase block (i.e. phase st), value=var_idx   \n",
    "\n",
    "phased_vcf_dic={}\n",
    "\n",
    "var_pos_list=[] # all blocks consequently \n",
    "header_lines_list=[] \n",
    "\n",
    "var_idx=0 # for the first variant, it's 1. 1-based indexing \n",
    "for line in vcf_file:\n",
    "    line_strip=line.strip() \n",
    "    if line_strip.startswith('#'):\n",
    "        header_lines_list.append(line_strip)\n",
    "        sample_names=line_strip.split('\\t')[9:11]# last line of header contains sample name\n",
    "    else:\n",
    "        var_idx+=1\n",
    "        line_parts=line_strip.split('\\t') \n",
    "        \n",
    "        gt_flags, sample1=line_parts[8:10] # sample of interest that is merged        \n",
    "        chrom=line_parts[0]\n",
    "        \n",
    "        if str(chrom_output)!=chrom:\n",
    "            print(chrom_output,chrom)\n",
    "        \n",
    "        var_pos=int(line_parts[1])\n",
    "        var_pos_list.append(var_pos)\n",
    "        #if var_pos < 7398830: #60780\n",
    "            \n",
    "        phased_vcf_dic[var_idx]=line_parts\n",
    "\n",
    "\n",
    "        sample1_split=sample1.split(\":\")\n",
    "        #gt_flags_split=gt_flags.split(\":\")\n",
    "        gt_id_flags=gt_flags.split(\":\").index(\"GT\")\n",
    "        ps_id_flags=gt_flags.split(\":\").index(\"PS\")\n",
    "        allele_sample1=sample1_split[gt_id_flags]        \n",
    "        if '.' not in allele_sample1:\n",
    "            block_id1=int(sample1_split[ps_id_flags])\n",
    "            if block_id1 in hap_blocks:\n",
    "                hap_blocks[block_id1].append(allele_sample1) # append new variants to the existing phase block\n",
    "                idc_blocks[block_id1].append(var_idx)\n",
    "            else:\n",
    "                hap_blocks[block_id1]=[allele_sample1] # creat new phase block\n",
    "                idc_blocks[block_id1]=[var_idx]\n",
    "\n",
    "        else:\n",
    "            print('There is a vriant that ONT is not phased.. removed it first')\n",
    "\n",
    "\n",
    "print('number of blocks in ont',len(hap_blocks))\n",
    "\n",
    "# print('ont')\n",
    "# for block_id, pos_block in pos_blocks.items():\n",
    "#     print(block_id,len(pos_block))\n",
    "# print('ont')\n",
    "for block_id1, hap_block  in hap_blocks.items():\n",
    "    print(block_id1,len(hap_block))\n",
    "hap_block[:2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'sim': [141700, 141716], 'dis': []}\n",
      "78486\n"
     ]
    }
   ],
   "source": [
    "#### Parsing pusod read\n",
    "# handling pop information reporting only once\n",
    "\n",
    "# reading short haplotype block file\n",
    "pseudo='data/'+str(chrom_output)+'/pairs.txt'\n",
    "file_hap = open(pseudo,'r'); \n",
    "pop_inf_dic={} # key variant index, value two lists \n",
    "for line in file_hap:\n",
    "    line_parts=line.strip().split('\\t')     # ['42081', '0', '42096', '0']\n",
    "    snp1_idx=int(line_parts[0])\n",
    "    allele1_snp1=int(line_parts[1])\n",
    "    snp2_idx=int(line_parts[2])\n",
    "    allele1_snp2=int(line_parts[3])\n",
    "    #pairs_list.append([snp1_idx,allele1_snp1,snp2_idx,allele1_snp2])\n",
    "    \n",
    "    snps_idx=[snp1_idx,snp2_idx]\n",
    "    snps_allele=[allele1_snp1,allele1_snp2]\n",
    "    host_idx=max(snps_idx)\n",
    "    guest_idx=min(snps_idx)\n",
    "    if host_idx not in pop_inf_dic.keys():\n",
    "        pop_inf_dic[host_idx]={'sim':list(),'dis':list()}\n",
    "    if allele1_snp1==allele1_snp2:\n",
    "        pop_inf_dic[host_idx]['sim'].append(guest_idx)  # sim shows the relation between two elements of a pop pair\n",
    "    else:\n",
    "        pop_inf_dic[host_idx]['dis'].append(guest_idx) \n",
    "\n",
    "# line_parts\n",
    "# print(line_parts,'\\n', pairs_list[0])\n",
    "# print(len(pairs_list))\n",
    "print(pop_inf_dic[host_idx])\n",
    "print(len(pop_inf_dic))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comparing one ONT block with all pop information\n",
    "\n",
    "# block_id1=8889839 # test for chr 19\n",
    "# hap_block=hap_blocks[block_id1]\n",
    "# idc_block=idc_blocks[block_id1]\n",
    "\n",
    "# if 1:\n",
    "def compare_ont_pop(hap_block,idc_block,pop_inf_dic,interval):\n",
    "    \n",
    "    # input: interval is based on var_i \n",
    "    # output: is ont_pop_block[var_idx]=ont_pop \n",
    "    \n",
    "    #var_i is the index of variant within block\n",
    "    #var_idx is the index of vriant globally in the VCF file     \n",
    "    \n",
    "    ont_pop_block={}\n",
    "    min_interval=min(list(interval))\n",
    "    max_interval=max(list(interval))\n",
    "    interval_idx=set(idc_block[min_interval:max_interval]) # set is faster than list for check if element is in it\n",
    "    for var_i in interval: # the ONT hap block\n",
    "        alleles=hap_block[var_i]\n",
    "        var_idx=idc_block[var_i]\n",
    "        \n",
    "        #var_i is the index of variant within block\n",
    "        #var_idx is the index of vriant globally in the VCF file \n",
    "        \n",
    "        \n",
    "        if var_idx in pop_inf_dic.keys():\n",
    "            list_pairs=pop_inf_dic[var_idx]\n",
    "            match_sim_list=[]\n",
    "            match_dis_list=[]\n",
    "            mismatch_sim_list=[]\n",
    "            mismatch_dis_list=[]\n",
    "            for sim_idx in pop_inf_dic[var_idx]['sim']: # sim shows the relation between two elements of a pop pair\n",
    "                \n",
    "                #if sim_idx in idc_block: # if the SNP  of pairs is in this ONT block\n",
    "                try:\n",
    "                    alleles_pop=hap_block[idc_block.index(sim_idx)] # extract the allele of the snp 'sim' which is the other snp in pair\n",
    "                    if sim_idx in interval_idx:\n",
    "                        if alleles_pop==alleles: \n",
    "                            match_sim_list.append(str(sim_idx))  \n",
    "                        else:\n",
    "                            mismatch_sim_list.append(str(sim_idx))     \n",
    "                except:\n",
    "                    pass\n",
    "                        \n",
    "            for sim_idx in pop_inf_dic[var_idx]['dis']: # sim shows the relation between two elements of a pop pair\n",
    "                #if sim_idx in idc_block: # if the SNP  of pairs is in this ONT block\n",
    "                try:\n",
    "                    alleles_pop=hap_block[idc_block.index(sim_idx)] # extract the allele of the snp 'sim' which is the other snp in pair\n",
    "                    if sim_idx in interval_idx:\n",
    "                        if alleles_pop==str(1-int(alleles[0]))+'|'+str(1-int(alleles[2])):  \n",
    "                            match_sim_list.append(str(sim_idx))  \n",
    "                        else:\n",
    "                            mismatch_sim_list.append(str(sim_idx))            \n",
    "                except:\n",
    "                    pass\n",
    "        \n",
    "            match_list=match_sim_list+match_dis_list\n",
    "            mismatch_list=mismatch_sim_list+mismatch_dis_list\n",
    "            ont_pop=[match_list,mismatch_list]\n",
    "        else:\n",
    "            ont_pop=[[],[]]\n",
    "\n",
    "        ont_pop_block[var_idx]=ont_pop\n",
    "    #print('comparison is done')\n",
    "    return  ont_pop_block\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# comparing ONT (hap_blocks_sample1_dic)with pop (short_blocks_dic_pos)\n",
    "\n",
    "ont_pop_blocks={}\n",
    "for block_id1, hap_block in hap_blocks.items(): # all ONT hap blocks  \n",
    "    idc_block=idc_blocks[block_id1]\n",
    "     \n",
    "    #if block_id1==27731803: #27731803: 60780 20588536 8889839\n",
    "    ont_pop_block=compare_ont_pop(hap_block,idc_block,pop_inf_dic,range(len(hap_block)))\n",
    "    ont_pop_blocks[block_id1]=ont_pop_block\n",
    "\n",
    "#hap_blocks.keys()\n",
    "#var_pos_list[31566-1] #grep -v \"#\" out.vcf | sed -n 31566,31566p\n",
    "#ont_pop_blocks[60780][31576]\n",
    "# print(ont_pop_blocks[60780][292])\n",
    "# print('for variant',var_pos_list[292-1],'we have',var_pos_list[268-1],var_pos_list[269-1],var_pos_list[261-1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cut candidate 4597794 5686 5687\n",
      "cut candidate 4601068 5690 5691\n",
      "flip candidate 9579 7346152\n",
      "cut candidate 12893528 16417 16418\n",
      "cut candidate 13263027 16944 16945\n",
      "flip candidate 17377 13562076\n",
      "flip candidate 17836 14169317\n",
      "flip candidate 20744 16734165\n",
      "flip candidate 23432 19293071\n",
      "cut candidate 19874459 2234 24344\n",
      "flip candidate 24379 20058500\n",
      "cut candidate 24604635 6286 28396\n",
      "flip candidate 28417 24624556\n",
      "cut candidate 30678263 3622 36737\n",
      "cut candidate 33843524 6463 39578\n",
      "cut candidate 48888592 972 58492\n",
      "cut candidate 53176928 1882 62893\n",
      "cut candidate 65836969 16461 77472\n",
      "cut candidate 67202386 17809 78820\n",
      "cut candidate 67225139 17817 78828\n",
      "flip candidate 81115 68697656\n",
      "cut candidate 72574945 23810 84821\n",
      "cut candidate 73731442 25059 86070\n"
     ]
    }
   ],
   "source": [
    "#aa=0\n",
    "\n",
    "\n",
    "#block_id1=27731803 # 60780  # 20588536 27731803 8889839\n",
    "#hap_block=hap_blocks[block_id1]\n",
    "\n",
    "#if 1:\n",
    "flip_list=[]\n",
    "cut_dic={}\n",
    "\n",
    "\n",
    "\n",
    "for block_id1, hap_block in hap_blocks.items():\n",
    "    cut_dic[block_id1]=[]\n",
    "\n",
    "    idc_block=idc_blocks[block_id1]\n",
    "    ont_pop_block=ont_pop_blocks[block_id1]\n",
    "\n",
    "    for var_i, var_idx in enumerate(idc_block): \n",
    "        #var_i is the index of variant within block\n",
    "        #var_idx is the index of vriant globally in the VCF file \n",
    "\n",
    "        ont_pop = ont_pop_block[var_idx]  # after a cut, it may use a newr version of this vsrisble\n",
    "        if ont_pop==[[],[]]:\n",
    "            out_pairs='.'\n",
    "        else:\n",
    "            [match_list,mismatch_list]=ont_pop    \n",
    "            len_mismatch_list=len(mismatch_list)\n",
    "            len_match_list=len(match_list)\n",
    "            condition=0\n",
    "            condition_flip=0\n",
    "            condition_cut=0\n",
    "            if len_mismatch_list>=2:\n",
    "                if len_match_list==0:\n",
    "                    condition=1\n",
    "                elif len_mismatch_list/len_match_list >= 0.75 :\n",
    "                    condition=1\n",
    "            if condition:\n",
    "                #print(len_mismatch_list,var_idx,var_pos_list[var_idx-1])\n",
    "                ont_pop_prev = ont_pop_block[var_idx-1] \n",
    "                ont_pop_nex = ont_pop_block[var_idx+1] \n",
    "\n",
    "                len_match_list_nex=len(ont_pop_nex[0])\n",
    "                len_mismatch_list_nex=len(ont_pop_nex[1])\n",
    "                len_match_list_prev=len(ont_pop_prev[0])\n",
    "                len_mismatch_list_prev=len(ont_pop_prev[1])\n",
    "\n",
    "                if len_match_list_nex>1.3*len_mismatch_list_nex and len_match_list_prev>1.3*len_mismatch_list_prev: #  ?? if no pop information in last var\n",
    "                    condition_flip=1\n",
    "                if len_match_list_prev==0:\n",
    "                    if  len_mismatch_list_nex>=2:\n",
    "                        condition_cut=1\n",
    "                elif len_mismatch_list_nex/len_match_list_prev >= 0.75:\n",
    "                    condition_cut=1\n",
    "\n",
    "            if condition_flip:\n",
    "                print('flip candidate', var_idx, var_pos_list[var_idx-1])\n",
    "                flip_list.append(var_idx)\n",
    "            if condition_cut and not condition_flip:\n",
    "                cut_pos=var_pos_list[var_idx-1] # cut_pos the starting position of new block\n",
    "                print('cut candidate', cut_pos,var_i,var_idx)   \n",
    "                cut_dic[block_id1].append(var_idx)\n",
    "                interval=range(var_i,len(hap_block))\n",
    "                \n",
    "\n",
    "                \n",
    "                \n",
    "                ont_pop_block=compare_ont_pop(hap_block,idc_block,pop_inf_dic,interval)\n",
    "#                 if cut_pos==43529026:\n",
    "#                     print('after')\n",
    "#                     print(ont_pop_block[49889])\n",
    "#                     print(ont_pop_block[49900])\n",
    "#                     break\n",
    "                    \n",
    "                    \n",
    "                    \n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# improving flips\n",
    "# for block_id1, hap_block in hap_blocks.items():\n",
    "#     idc_block=idc_blocks[block_id1]\n",
    "#     ont_pop_block=ont_pop_blocks[block_id1]\n",
    "\n",
    "phased_vcf_dic_improved=phased_vcf_dic\n",
    "for var_idx in flip_list:\n",
    "    line_parts=phased_vcf_dic[var_idx]\n",
    "    gt_flags, sample1=line_parts[8:10]\n",
    "    sample1_split=sample1.split(\":\")\n",
    "    gt_id_flags=gt_flags.split(\":\").index(\"GT\")        \n",
    "    allele_sample1=sample1_split[gt_id_flags]   \n",
    "    \n",
    "    allele_sample1_flp=str(1-int(allele_sample1[0]))+'|'+str(1-int(allele_sample1[2]))\n",
    "    sample1_split_flp=sample1_split\n",
    "    sample1_split_flp[gt_id_flags]=allele_sample1_flp\n",
    "    line_parts_flp=line_parts\n",
    "    line_parts_flp[9]=':'.join(sample1_split_flp)\n",
    "    phased_vcf_dic_improved[var_idx]=line_parts_flp\n",
    "    #block_id1=int(sample1_split[ps_id_flags])\n",
    "    #ps_id_flags=gt_flags.split(\":\").index(\"PS\")\n",
    "\n",
    "#changing block is done in reporting \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for block_id1, hap_block in hap_blocks.items():\n",
    "\n",
    "#     idc_block=idc_blocks[block_id1]\n",
    "#     ont_pop_block=ont_pop_blocks[block_id1]\n",
    "\n",
    "#     for var_i, var_idx in enumerate(idc_block): \n",
    "\n",
    "#             cut_block=cut_dic[block_id1]\n",
    "\n",
    "\n",
    "#                 new_blockid=cut_pos\n",
    "#                 dic1{}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# block_id1=60780\n",
    "# new_blockid_dic={}\n",
    "# cut_block=cut_dic[block_id1]\n",
    "# cut_block_complete=[idc_block[0]]+cut_block+[idc_block[-1]] # including start and end var_idx of block\n",
    "# cut_block_complete\n",
    "# cut_dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71453\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# creating new block id dic\n",
    "\n",
    "new_blockid_dic={}\n",
    "\n",
    "for block_id1 in hap_blocks.keys():\n",
    "\n",
    "    cut_block=cut_dic[block_id1]\n",
    "    idc_block=idc_blocks[block_id1]\n",
    "\n",
    "    cut_block_complete=[idc_block[0]]+cut_block+[idc_block[-1]] # including start and end var_idx of block\n",
    "\n",
    "    for i in range(1,len(cut_block_complete)):\n",
    "        pre_cut_idx=cut_block_complete[i-1]\n",
    "        next_cut_idx=cut_block_complete[i]\n",
    "\n",
    "        blockid=var_pos_list[pre_cut_idx-1]\n",
    "        for k in range(pre_cut_idx,next_cut_idx+1):\n",
    "            new_blockid_dic[k]=blockid\n",
    "\n",
    "\n",
    "\n",
    "print(len(new_blockid_dic))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 71453\n",
      "60780\n",
      "7398833\n",
      "8737935\n",
      "8889839\n",
      "20588536\n",
      "27731803\n"
     ]
    }
   ],
   "source": [
    "a=sorted(new_blockid_dic.keys())\n",
    "print(a[0],a[-1])\n",
    "\n",
    "\n",
    "for block_id1 in hap_blocks.keys():\n",
    "    print(block_id1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dic1={1:1}\n",
    "\n",
    "\n",
    "phased_vcf_dic_improved2=phased_vcf_dic_improved\n",
    "\n",
    "sorted_var_indx=sorted(phased_vcf_dic_improved.keys())\n",
    "\n",
    "cut_dic_remain=cut_dic\n",
    "\n",
    "for var_indx in sorted_var_indx: # imprvd_phased_vcf_dic\n",
    "    \n",
    "    #if var_indx in dic1: # those variant that we need to change their block id \n",
    "    line_parts=phased_vcf_dic_improved[var_indx]\n",
    "    gt_flags, sample1=line_parts[8:10]\n",
    "    sample1_split=sample1.split(\":\")\n",
    "    ps_id_flags=gt_flags.split(\":\").index(\"PS\")\n",
    "    block_id1=int(sample1_split[ps_id_flags])\n",
    "\n",
    "    new_blockid=str(new_blockid_dic[var_indx])\n",
    "\n",
    "\n",
    "    #cut_dic[block_id1]\n",
    "    sample1_split_changed=sample1.split(\":\")\n",
    "    sample1_split_changed[ps_id_flags]=new_blockid\n",
    "    line_parts_changed=line_parts\n",
    "    line_parts_changed[9]=':'.join(sample1_split_changed)\n",
    "    phased_vcf_dic_improved2[var_indx]=line_parts_changed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# max(phased_vcf_dic_improved.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# line_parts=phased_vcf_dic[10]\n",
    "# gt_flags, sample1=line_parts[8:10]\n",
    "# sample1_split=sample1.split(\":\")\n",
    "# ps_id_flags=gt_flags.split(\":\").index(\"PS\")\n",
    "# block_id1=int(sample1_split[ps_id_flags])\n",
    "# block_id1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phased_vcf_dic_improved=phased_vcf_dic\n",
    "# for var_idx in cut_set:\n",
    "#     line_parts=phased_vcf_dic[var_idx]\n",
    "#     gt_flags, sample1=line_parts[8:10]\n",
    "#     sample1_split=sample1.split(\":\")        \n",
    "#     gt_id_flags=gt_flags.split(\":\").index(\"GT\")        \n",
    "#     allele_sample1=sample1_split[gt_id_flags]   \n",
    "#     block_id1=int(sample1_split[ps_id_flags])\n",
    "#     ps_id_flags=gt_flags.split(\":\").index(\"PS\")\n",
    "    \n",
    "    \n",
    "    \n",
    "#     allele_sample1_flp=str(1-int(allele_sample1[0]))+'|'+str(1-int(allele_sample1[2]))\n",
    "#     sample1_split_flp=sample1_split\n",
    "#     sample1_split_flp[gt_id_flags]=allele_sample1_flp\n",
    "#     line_parts_flp=line_parts\n",
    "#     line_parts_flp[9]=':'.join(sample1_split_flp)\n",
    "#     phased_vcf_dic_improved[var_idx]=line_parts_flp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "vcf_file_improved_address = 'data/'+str(chrom_output)+'/out_improved.vcf' #tst.vcf' #\n",
    "\n",
    "# reading phased VCF file,  use grep before that, only 0|1 or 1|0. bi -allelic\n",
    "vcf_file_improved = open(vcf_file_improved_address,'w');  # phased_vcf_dic\n",
    "\n",
    "sorted_var_indx=sorted(phased_vcf_dic_improved2.keys())\n",
    "\n",
    "for header_line in header_lines_list:\n",
    "    vcf_file_improved.write(header_line+'\\n')\n",
    "for var_indx in sorted_var_indx: # imprvd_phased_vcf_dic\n",
    "    line_parts=phased_vcf_dic_improved2[var_indx]\n",
    "    vcf_file_improved.write('\\t'.join(line_parts)+'\\n')\n",
    "vcf_file_improved.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # summary\n",
    "# import numpy as np\n",
    "\n",
    "# len_list=[]\n",
    "# lengths=[]\n",
    "# for block_id1, idc_block  in idc_blocks.items():\n",
    "#     indc_srtd=sorted(idc_block)\n",
    "#     diff=var_pos_list[indc_srtd[-1]-1]-var_pos_list[indc_srtd[0]-1] # all of the variant in a block are consequtive \n",
    "#     lengths.append(diff)\n",
    "\n",
    "#     len_list.append(len(hap_block))\n",
    "\n",
    "# all_len=sorted(lengths, reverse=True)\n",
    "# csum=np.cumsum(all_len)\n",
    "# n2=int(sum(lengths)/2)\n",
    "# csumn2=min(csum[csum >= n2])\n",
    "# ind=np.where(csum == csumn2)\n",
    "\n",
    "# n50 = all_len[int(ind[0])]\n",
    "\n",
    "# print('** Chr 19 ** ')\n",
    "# print('Number phased heterzygous SNVs:')\n",
    "# print('Number non-phased heterzygous SNVs, 0/0:','1/1:')\n",
    "# print('Number homozygous SNVs:')\n",
    "\n",
    "# print('N50 of phase blocks: ',n50)\n",
    "# #print('Mean of  genomic length of haplotype 'np.mean(lengths))\n",
    "# print('Mean number of SNV in phase block ',np.mean(len_list))\n",
    "# print('Number conflict SNV with population information: ','%')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
