{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "from sys import argv\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "def read_vcf_file_with_truth(vcf_file_address):    \n",
    "    \n",
    "    \"\"\"\n",
    "    Reading the vcf file\n",
    "    \n",
    "    input: vcf file\n",
    "    outputs: \n",
    "            lines_list: list of string. each string is a line of phased vcf file.\n",
    "            var_pos_het_list: genomic position of phased hetrozygous variants in the vcf file. \n",
    "            line_number_het_list: list of line numbers in the vcf file that are phased hetrozygous variant (needed in phasing)\n",
    "            id_blocks: list of ids of phas blocks\n",
    "            allele_blocks: list of list\n",
    "            var_pos_blocks: list of list\n",
    "            stats_vcf: [homozygous0_num, homozygous1_num, hetrozygous_nonphased, hetrozygous_phased, genomic_length_blocks, n50]\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    allele_truth_dic = {}\n",
    "    \n",
    "    vcf_file = open(vcf_file_address,'r')\n",
    "\n",
    "    lines_list=[]                  #  lines  of phased vcf  needed for reporting improved VCF \n",
    "    var_pos_het_list=[]                # position of phased hetrozygous variants all blocks consequently. \n",
    "\n",
    "    # The followings are for phased hetrozygous variants.\n",
    "    id_blocks = []                 # list of list. Outer list corresponds to phase block. Inner list contains block_id\n",
    "    \n",
    "    allele_blocks = []             # list of list. Outer list corresponds to phase block. Inner list contains alleles of hetro variants \n",
    "    var_pos_blocks = []            # list of list. Outer list corresponds to phase block. Inner list contains genomic positions of hetro variants   \n",
    "\n",
    "    \n",
    "    line_number_het_list = []            # line number of phased hetrozygous variant. We need it for reporting improved version\n",
    "    lines_list = []\n",
    "    \n",
    "    first_het_variant = True\n",
    "    line_number = 0\n",
    "    \n",
    "    homozygous0_num = 0\n",
    "    homozygous1_num = 0\n",
    "    hetrozygous_nonphased = 0\n",
    "    hetrozygous_phased = 0\n",
    "    \n",
    "    \n",
    "    for line in vcf_file:\n",
    "        \n",
    "        line_number += 1\n",
    "        \n",
    "        line_strip = line.strip() \n",
    "        \n",
    "        lines_list.append(line_strip)\n",
    "        \n",
    "        if line_strip.startswith('#'):\n",
    "            pass \n",
    "            #header_lines_list.append(line_strip)\n",
    "            #sample_names = line_strip.split('\\t')[9:11]            # last line of header contains sample name\n",
    "        else:\n",
    "\n",
    "            line_parts=line_strip.split('\\t') \n",
    "\n",
    "            chrom = line_parts[0]\n",
    "\n",
    "#             if str(chrom_output)!=chrom:\n",
    "#                 print(chrom_output,chrom)\n",
    "\n",
    "            var_pos = int(line_parts[1])                           # genomic position of variants\n",
    "            \n",
    "\n",
    "            \n",
    "            format_genotype, values_genotype = line_parts[8:10]    # 'GT:GQ:DP:AF:GL:PS', '0|1:255:.:.:.,0,.:60780'\n",
    "            \n",
    "            values_genotype_splitted = values_genotype.split(':')\n",
    "            format_genotype_splitted = format_genotype.split(':')\n",
    "            \n",
    "            gt_index = format_genotype_splitted.index(\"GT\")           #  index of allele in  values_genotype \n",
    "            \n",
    "            allele = values_genotype_splitted[gt_index]\n",
    "            \n",
    "            \n",
    "            # how should we handle '2' in allele ?  \n",
    "            \n",
    "#             if './.' in allele:\n",
    "#                 print(\"There is a vriant with genomic position \"+str(var_pos)+\" that is not genotyped. Remove it first.\")\n",
    "#                 exit(1)\n",
    "                \n",
    "            \n",
    "            # if '/' in allele: print(\"There is a vriant with genomic position \"+str(var_pos)+\" that is not phased. Remove it first.\")            \n",
    "\n",
    "            if allele == '0/1' or allele == '1/0':\n",
    "                hetrozygous_nonphased += 1\n",
    "\n",
    "            if allele == '0|0' or allele == '0/0':\n",
    "                homozygous0_num += 1\n",
    "                \n",
    "            if allele == '1|1' or allele == '1/1':\n",
    "                homozygous1_num += 1\n",
    "                    \n",
    "            if (allele == '0|1' or allele == '1|0'):\n",
    "            \n",
    "                hetrozygous_phased += 1\n",
    "                \n",
    "                var_pos_het_list.append(var_pos)\n",
    "                \n",
    "                line_number_het_list.append(line_number)\n",
    "                \n",
    "                ps_index = format_genotype_splitted.index(\"PS\")           #  index of phase set in values_genotype \n",
    "                id_block = values_genotype_splitted[ps_index]\n",
    "                \n",
    "                \n",
    "                \n",
    "                if first_het_variant:           # for the first het variant\n",
    "                    first_het_variant = False\n",
    "                    \n",
    "                    allele_block = [allele]\n",
    "                    var_pos_block = [int(var_pos)]\n",
    "                    id_blocks.append(id_block)\n",
    "                    \n",
    "                    \n",
    "                else:                              # for the rest of het variants\n",
    "                    if id_block in id_blocks:\n",
    "                        allele_block.append(allele)\n",
    "                        var_pos_block.append(int(var_pos)) \n",
    "\n",
    "                    else:\n",
    "\n",
    "                        # add previous block to the list of all blocks\n",
    "                        allele_blocks.append(allele_block)\n",
    "                        var_pos_blocks.append(var_pos_block)\n",
    "\n",
    "                        # creat new phase block\n",
    "                        allele_block = [allele]\n",
    "                        var_pos_block = [int(var_pos)]\n",
    "                        id_blocks.append(id_block)\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "                values_genotype_truth = line_parts[10]\n",
    "                values_genotype_truth_splitted = values_genotype_truth.split(':')\n",
    "                allele_truth = values_genotype_truth_splitted[gt_index]\n",
    "                id_block_truth = values_genotype_truth_splitted[ps_index]\n",
    "                #if allele_truth != './.':\n",
    "                if allele_truth == '0|1' or allele_truth == '1|0':\n",
    "                    allele_truth_dic[var_pos]= str(id_block_truth)+':'+str(allele_truth)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    # # for the last het variant, we  finish the last block.\n",
    "    allele_blocks.append(allele_block)\n",
    "    var_pos_blocks.append(var_pos_block)\n",
    "    \n",
    "    \n",
    "          \n",
    "    genomic_length_blocks = [] \n",
    "    \n",
    "    for var_pos_block in var_pos_blocks:\n",
    "        genomic_length_blocks.append(var_pos_block[-1]-var_pos_block[0])\n",
    "        \n",
    "    values_sorted = sorted(genomic_length_blocks, reverse=True)\n",
    "    csum = np.cumsum(values_sorted)\n",
    "    \n",
    "    n2 = int(sum(values_sorted)/2)\n",
    "    csumn2 = min(csum[csum >= n2])\n",
    "    ind = np.where(csum == csumn2)\n",
    "    n50 = values_sorted[int(ind[0])]\n",
    "\n",
    "        \n",
    "    \n",
    "    stats_vcf = [homozygous0_num, homozygous1_num, hetrozygous_nonphased, hetrozygous_phased, genomic_length_blocks, n50]\n",
    "\n",
    "\n",
    "\n",
    "    return lines_list, var_pos_het_list, line_number_het_list, id_blocks, allele_blocks, var_pos_blocks, stats_vcf, allele_truth_dic\n",
    "\n",
    "\n",
    "def read_file_pairs_forward(file_pairs_address):\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    Reading the pairs.txt file and save it in a dictinary only once.\n",
    "\n",
    "    input: pairs.txt\n",
    "    Each line of input file has three part \n",
    "        position of var1, position of  var2, realtion between phasing of var1 and that of var2\n",
    "        10 20 identical\n",
    "        \n",
    "        Each pair reports only one in output dictionary. We report the example for the key   20:[[10],[]]\n",
    "        Becuase, in phase block we go variant by vairant. So the phasing of previous variants are important not later's\n",
    "        \n",
    "    output: a dictinary, key: genomic position of variant, value: [[],[]] \n",
    "    \n",
    "            first list: the genomic position of variant with identical phasing \n",
    "            second list: the genomic position of variant with opposite phasing\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    file_pairs = open(file_pairs_address,'r'); \n",
    "    \n",
    "    pop_inf_dic = {}                              # key variant index, value two lists \n",
    "    for line in file_pairs:\n",
    "        \n",
    "        line_parts = line.strip().split('\\t')     # ['42081', '0', '42096', '0']\n",
    "\n",
    "        \n",
    "        \n",
    "        snv1_pos = int(line_parts[0]) \n",
    "        snv2_pos = int(line_parts[1])                \n",
    "\n",
    "        relation_phasing = line_parts[2]\n",
    "        \n",
    "        snvs_pos = [snv1_pos, snv2_pos]\n",
    "        host_pos = max(snvs_pos)\n",
    "        guest_pos = min(snvs_pos)\n",
    "        \n",
    "        if host_pos not in pop_inf_dic.keys():\n",
    "            \n",
    "            # first list identical phase\n",
    "            # second list opposite phase\n",
    "            pop_inf_dic[host_pos] = [[], []]\n",
    "            \n",
    "        if relation_phasing == 'identical':\n",
    "            pop_inf_dic[host_pos][0].append(guest_pos)  \n",
    "        elif relation_phasing == 'opposite':\n",
    "            pop_inf_dic[host_pos][1].append(guest_pos) \n",
    "            \n",
    "    return pop_inf_dic  # [vars_identical_phase, vars_opposite_phase]=pop_inf_dic[var_pos]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def read_file_pairs_forward_backward(file_pairs_address):\n",
    "\n",
    "    \"\"\"\n",
    "    see read_file_pairs_forward\n",
    "    \n",
    "    The only difference is that a pair is reported twic. \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    file_pairs = open(file_pairs_address,'r'); \n",
    "    \n",
    "    pop_inf_dic = {}                              # key variant index, value two lists \n",
    "    for line in file_pairs:\n",
    "        \n",
    "        line_parts = line.strip().split('\\t')     # ['42081', '0', '42096', '0']\n",
    "\n",
    "         \n",
    "        snv1_pos = int(line_parts[0]) \n",
    "        snv2_pos = int(line_parts[1])                \n",
    "\n",
    "        relation_phasing = line_parts[2]\n",
    "        \n",
    "        host_pos = snv1_pos\n",
    "        guest_pos = snv2_pos\n",
    "        \n",
    "        if host_pos not in pop_inf_dic.keys():\n",
    "            \n",
    "            # first list identical phase\n",
    "            # second list opposite phase\n",
    "            pop_inf_dic[host_pos] = [[], []]\n",
    "            \n",
    "        if relation_phasing == 'identical':\n",
    "            pop_inf_dic[host_pos][0].append(guest_pos)  \n",
    "        elif relation_phasing == 'opposite':\n",
    "            pop_inf_dic[host_pos][1].append(guest_pos) \n",
    "   \n",
    "            \n",
    "        host_pos = snv2_pos\n",
    "        guest_pos = snv1_pos\n",
    "        \n",
    "        if host_pos not in pop_inf_dic.keys():\n",
    "            \n",
    "            # first list identical phase\n",
    "            # second list opposite phase\n",
    "            pop_inf_dic[host_pos] = [[], []]\n",
    "            \n",
    "        if relation_phasing == 'identical':\n",
    "            pop_inf_dic[host_pos][0].append(guest_pos)  \n",
    "        elif relation_phasing == 'opposite':\n",
    "            pop_inf_dic[host_pos][1].append(guest_pos) \n",
    "            \n",
    "    return pop_inf_dic  # [vars_identical_phase, vars_opposite_phase]=pop_inf_dic[var_pos]\n",
    "\n",
    "      \n",
    "\n",
    "\n",
    "\n",
    "def compare_phase_block_pop(allele_block, var_pos_block, pop_inf_dic, lower_bound, upper_bound):\n",
    "\n",
    "    \n",
    "    \"\"\"\n",
    "    Compare alleles of phased VCF with pairs (from the population information)\n",
    "    \n",
    "    input:  allele_block, var_pos_block    (phased VCF)\n",
    "            pop_inf_dic \n",
    "    \n",
    "    output: a list of [[],[]] \n",
    "    \n",
    "                        Inner list correspond to a variant (host). \n",
    "                        first inner list: positons of variants that are matched with the host variant (based on population information)\n",
    "                                         the status of match can be either identical or opposite phasing.                \n",
    "                        second inner list mismatched\n",
    "    \n",
    "    \n",
    "    Exmpale 1:  pair from population information:  10 20 identical\n",
    "                if phased vcf   10 0|1    20 0|1, the 10 and 20 are matched.\n",
    "                if phased vcf   10 0|1    20 1|0, the 10 and 20 are mismatched.\n",
    "\n",
    "    Exmpale 2:  pair from population information:  10 20 opposite\n",
    "                if phased vcf   10 0|1    20 0|1, the 10 and 20 are mismatched.\n",
    "                if phased vcf   10 0|1    20 1|0, the 10 and 20 are matched.\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    # the results of comparison between \n",
    "    \n",
    "    comparison_result_block = []\n",
    "    for var_i in range(len(allele_block)):\n",
    "\n",
    "        #var_i is the index of variant within the block of phased vcf\n",
    "\n",
    "        var_pos = var_pos_block[var_i]\n",
    "        allele = allele_block[var_i]\n",
    "\n",
    "        if var_pos >= lower_bound and  var_pos <= upper_bound:\n",
    "\n",
    "            #alleles=hap_block[var_i]\n",
    "            #var_idx=idc_block[var_i]\n",
    "            #var_i is the index of variant within block\n",
    "            #var_idx is the index of vriant globally in the VCF file \n",
    "            if var_pos in pop_inf_dic.keys():\n",
    "    \n",
    "                [vars_identical_phase, vars_opposite_phase] = pop_inf_dic[var_pos]  \n",
    "        \n",
    "                \n",
    "                matched_identical_phase_list = []\n",
    "                mismatched_identical_phase_list = []\n",
    "                matched_opposite_phase_list = []\n",
    "                mismatched_opposite_phase_list = []\n",
    "                #for sim_idx in vars_identical_phase: # sim shows the relation between two elements of a pop pair\n",
    "                \n",
    "                \n",
    "                # The differene between the two following for is the comparing the allele_var_guest and  allele (flliped)\n",
    "                \n",
    "                for var_pos_identical_phase in vars_identical_phase: # Those variant that have the same phasing with var_pos\n",
    "\n",
    "                    if var_pos_identical_phase >= lower_bound and var_pos_identical_phase <= upper_bound:\n",
    "                        try:         # if the SNP  of pairs is in this block of phased VCF\n",
    "                            \n",
    "                            allele_var_guest= allele_block[var_pos_block.index(var_pos_identical_phase)]\n",
    "                                                   \n",
    "                            if allele_var_guest == allele: \n",
    "                                matched_identical_phase_list.append(var_pos_identical_phase)\n",
    "                            else:\n",
    "                                mismatched_identical_phase_list.append(var_pos_identical_phase) \n",
    "\n",
    "                        except:\n",
    "                            pass\n",
    "\n",
    "                    \n",
    "                for var_pos_opposite_phase in vars_opposite_phase: # Those variant that have the same phasing with var_pos\n",
    "                    \n",
    "                    if var_pos_opposite_phase >= lower_bound and  var_pos_opposite_phase <= upper_bound:\n",
    "                        try:         # if the SNP  of pairs is in this block of phased VCF        \n",
    "                            allele_var_guest = allele_block[var_pos_block.index(var_pos_opposite_phase)]\n",
    "                                                \n",
    "                            if allele_var_guest == str(1-int(allele[0]))+'|'+str(1-int(allele[2])):                   \n",
    "                                matched_opposite_phase_list.append(var_pos_opposite_phase)\n",
    "                            else:\n",
    "                                mismatched_opposite_phase_list.append(var_pos_opposite_phase) \n",
    "                        except:\n",
    "                            pass                \n",
    "                    \n",
    "                matched_list = matched_identical_phase_list + matched_opposite_phase_list\n",
    "                mismatched_list = mismatched_identical_phase_list + mismatched_opposite_phase_list\n",
    "                \n",
    "                comparison_result = [matched_list, mismatched_list]\n",
    "            else:\n",
    "                comparison_result = [[],[]]\n",
    "        else:\n",
    "            comparison_result = [[],[]]\n",
    "            \n",
    "        comparison_result_block.append(comparison_result)\n",
    "                \n",
    "        #print('comparison is done')\n",
    "    return  comparison_result_block\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def report_comparison_withtruth(report_out_address, comparison_result_blocks, chrom, allele_truth_dic):\n",
    "\n",
    "    file_report= open(report_out_address,'w'); \n",
    "    file_report.write(\"#chr\\t var_pos\\t blockid:alleles_ont \\t mismatched_pair \\t  matched_pair  \\n\")\n",
    "    \n",
    "    qual_blocks=[]\n",
    "    \n",
    "    for block_i, block_id  in enumerate(id_blocks):\n",
    "        \n",
    "        allele_block = allele_blocks[block_i]\n",
    "        var_pos_block = var_pos_blocks[block_i]\n",
    "        \n",
    "                \n",
    "    \n",
    "        file_report.write('\\t'.join([\"#Phase Block with ID  \",str(block_id)])+\"\\n\")\n",
    "\n",
    "        comparison_result_block = comparison_result_blocks[block_i]\n",
    "        \n",
    "        qual_block=[]\n",
    "        \n",
    "        \n",
    "        for var_i in range(len(allele_block)): # var_i internal index of this phase block\n",
    "\n",
    "            allele=allele_block[var_i]\n",
    "            var_pos=var_pos_block[var_i]\n",
    "\n",
    "\n",
    "            list_write=[]\n",
    "\n",
    "            \n",
    "            if var_pos in allele_truth_dic:\n",
    "                list_write.append(str(allele_truth_dic[var_pos]))\n",
    "                \n",
    "                allele_truth= allele_truth_dic[var_pos][-3:]\n",
    "                allele_truth_revert = str(1-int(allele_truth[0]))+'|'+str(1-int(allele_truth[2]))\n",
    "                \n",
    "                if allele == allele_truth_dic[var_pos][-3:]:\n",
    "                    sample_vs_truth = 'same'\n",
    "                elif allele == allele_truth_revert:\n",
    "                    sample_vs_truth = 'flipped'\n",
    "                    \n",
    "                list_write.append(sample_vs_truth)\n",
    "                \n",
    "                \n",
    "            else:\n",
    "                list_write.append('.')\n",
    "                list_write.append('.')\n",
    "                \n",
    "                \n",
    "                \n",
    "            \n",
    "            list_write.append(str(chrom))\n",
    "            list_write.append(str(var_pos))\n",
    "            \n",
    "            list_write.append(str(block_id)+':'+str(allele))\n",
    "\n",
    "            [matched_list, mismatched_list]=comparison_result_block[var_i]\n",
    "            \n",
    "            if len(matched_list)+len(mismatched_list):                # if we have population information\n",
    "                if len(matched_list):\n",
    "                    qual = len(matched_list)/(len(mismatched_list)+len(matched_list))\n",
    "                else:\n",
    "                    qual = 0\n",
    "                qual_block.append(qual)\n",
    "                \n",
    "                \n",
    "            if len(mismatched_list):\n",
    "                list_write.append(','.join([str(i) for i in mismatched_list]))\n",
    "            else: \n",
    "                list_write.append('.')\n",
    "\n",
    "            if len(matched_list):\n",
    "                list_write.append(','.join([str(i) for i in matched_list]))\n",
    "            else: \n",
    "                list_write.append('.')\n",
    "\n",
    "            file_report.write('\\t'.join(list_write)+\"\\n\")\n",
    "        \n",
    "        \n",
    "        qual_blocks.append(qual_block)\n",
    "        \n",
    "        #file_report.write(\"\\n\") # new  block\n",
    "\n",
    "    file_report.close()\n",
    "\n",
    "    return qual_blocks\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def report_qc(report_qc_address, id_blocks, qual_blocks, allele_blocks, stats_vcf):\n",
    "    \n",
    "    [homozygous0_num, homozygous1_num, hetrozygous_nonphased, hetrozygous_phased, genomic_length_blocks, n50] = stats_vcf\n",
    "    \n",
    "    file_report_qc= open(report_qc_address, 'w'); \n",
    "    file_report_qc.write(\"Quality report for chromosome \"+str(chrom)+\":\\n \\n\")\n",
    "    \n",
    "    file_report_qc.write(\"Number of homozygous variants is \"+str(homozygous0_num+homozygous1_num)+\", including \"+str(homozygous0_num)+\" variants with 0|0 and \"+str(homozygous1_num)+\" variants with 1|1:.\\n\")\n",
    "    file_report_qc.write(\"Number of non-phased heterozygous variants is \"+str(hetrozygous_nonphased)+\".\\n\")\n",
    "    file_report_qc.write(\"Number of heterozygous variants in all blocks is \"+str(hetrozygous_phased)+\".\\n\\n\")\n",
    "    \n",
    "    file_report_qc.write(\"N50 of phase blocks is \"+str(n50)+\"bp.\\n\\n\")\n",
    "    \n",
    "    for block_i, block_id  in enumerate(id_blocks):\n",
    "        \n",
    "        qual_block=qual_blocks[block_i]\n",
    "        allele_block=allele_blocks[block_i]\n",
    "        genomic_length_block = genomic_length_blocks[block_i]\n",
    "        \n",
    "        \n",
    "        file_report_qc.write(\"Phase block \"+str(block_i)+\":\\n\")\n",
    "        \n",
    "        \n",
    "        file_report_qc.write(\"The phase block spans \"+str(genomic_length_block)+\" bases.\\n\")\n",
    "        file_report_qc.write(\"The genomic position of first variant of the phase block is \"+str(block_id)+\".\\n\")\n",
    "        \n",
    "        if len(qual_block):\n",
    "            \n",
    "            q=np.mean(qual_block)\n",
    "            file_report_qc.write(\"Quality of block is \"+str(round(q,5))+\".\\n\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            file_report_qc.write(\"Quality of block: There is no population information for this block.\\n\")\n",
    "            \n",
    "            \n",
    "        file_report_qc.write(\"Number of variants in the phase block is \"+str(len(allele_block))+\".\\n\")\n",
    "        \n",
    "\n",
    "        file_report_qc.write(\"\\n\") # new  block\n",
    "\n",
    "    file_report_qc.close()\n",
    "    return 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    \n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    Input: phased vcf\n",
    "           pairs.txt\n",
    "           \n",
    "    \n",
    "    Output: report qc\n",
    "            report match and mismatches\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    type_data ='ont'\n",
    "    chrom = 22\n",
    "\n",
    "    vcf_file_address = type_data+'/input/'+str(chrom)+'_'+type_data+'_true.vcf'  # truth\n",
    "    #vcf_file_address = argv[1]\n",
    "    \n",
    "    lines_list, var_pos_het_list, line_number_het_list, id_blocks, allele_blocks, var_pos_blocks, stats_vcf, allele_truth_dic = read_vcf_file_with_truth(vcf_file_address)\n",
    "\n",
    "\n",
    "    file_pairs_address= type_data+'/input/'+str(chrom)+'_'+type_data+'_pairs_500_0.98.txt'\n",
    "    #file_pairs_address= argv[2]\n",
    "    \n",
    "    \n",
    "    \n",
    "    for Forward_or_both in  ['both', 'forward']:\n",
    "\n",
    "        if Forward_or_both == 'forward':\n",
    "            pop_inf_dic = read_file_pairs_forward(file_pairs_address)         \n",
    "\n",
    "        if Forward_or_both == 'both':                       # reproting pairs both forward and backward\n",
    "            pop_inf_dic = read_file_pairs_forward_backward(file_pairs_address)      \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # comparing input phased vcf (hap_blocks_sample1_dic) with pop (short_blocks_dic_pos)\n",
    "        comparison_result_blocks=[]\n",
    "        for block_i in range(len(id_blocks)):\n",
    "\n",
    "            allele_block = allele_blocks[block_i]\n",
    "            var_pos_block = var_pos_blocks[block_i]\n",
    "\n",
    "            # lower_bound and upper_bound is used for limit the span of comparison.\n",
    "            # Here, there is no limit. So, we set it from the begining to end of genomic position of block.\n",
    "            # The limit is used for updating the comparison result after cutting the block (in improver.py).\n",
    "            lower_bound = var_pos_block[0]         \n",
    "            upper_bound = var_pos_block[-1]\n",
    "\n",
    "            comparison_result_block = compare_phase_block_pop(allele_block, var_pos_block, pop_inf_dic, lower_bound, upper_bound)\n",
    "            comparison_result_blocks.append(comparison_result_block)\n",
    "\n",
    "\n",
    "\n",
    "        report_out_address=''+str(chrom)+'_report_mismatches_'+Forward_or_both+'.txt'\n",
    "        qual_blocks = report_comparison_withtruth(report_out_address, comparison_result_blocks, chrom, allele_truth_dic)\n",
    "\n",
    "        report_qc_address=''+str(chrom)+'_report_qc_'+Forward_or_both+'.txt'\n",
    "        report_qc(report_qc_address, id_blocks, qual_blocks, allele_blocks, stats_vcf)\n",
    "\n",
    "  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
